### PR TITLE
Remove references of deprecated const IL Opcode

### PR DESCRIPTION
**Pull request by wbh123456**
Remove all references to the IL Opcodes in the `const` category listed in #2657, namely,  `buconst`, `iuconst`, `luconst`, `cconst`. 
Also removed `TR::Node::cconst` 

*Anything in `SwitchAnalyzer.cpp` as they will be removed explicitly in #3983*

Issue: #2657 
Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

### CODE REVIEW

**Comment by andrewcraik** [comment]
Before removal of opcodes from optimizations happens we need to ensure no downstream projects are generating any instances of the opcode. I would go so far as to suggest that the tree verifier that runs during compilation (the old one as well as the new non-default one) are both modified to fatal assert if the opcode is seen once we believe all sources of the opcode have been eliminated. That way no one will add any code to generate these deprecated opcodes and removal can progress. Without some kind of always on assert and verification of downstream consumers I worry we may be adding bugs (either functional or performance) by removing handling...

**Comment by 0xdaryl** [comment]
Please check the linter failures.

File: compiler/x/codegen/BinaryEvaluator.cpp [hunk 1]
```diff
@@ -1784,8 +1783,7 @@ TR::Register *OMR::X86::TreeEvaluator::csubEvaluator(TR::Node *node, TR::CodeGen
       }
 
    TR::Register * testRegister = secondChild->getRegister();
-   if (secondChild->getOpCodeValue() == TR::cconst &&
```
Comment (wbh123456) [inline compiler/x/codegen/BinaryEvaluator.cpp]: This whole if statement can be removed.  The condition is always `false` as the `TR::cconst` no longer exists

File: compiler/x/codegen/BinaryEvaluator.cpp [hunk 2]
```diff
@@ -1252,8 +1252,7 @@ TR::Register *OMR::X86::TreeEvaluator::caddEvaluator(TR::Node *node, TR::CodeGen
...
```
Comment (wbh123456) [inline compiler/x/codegen/BinaryEvaluator.cpp]: This whole if statement can be removed. The condition is always false as the TR::cconst no longer exists

**Comment by wbh123456** [comment]
@vijaysun-omr This PR is good for a review.

**Comment by vijaysun-omr** [comment]
@genie-omr build all

**Comment by wbh123456** [comment]
> @genie-omr build all

I suspect the build failure was caused by the linking error of `cconstEvaluator` because I removed the definition of it but did not remove its declaration. In the new fore-push, I recovered the definition of `cconstEvaluator`. Could you build this PR again?

**Comment by vijaysun-omr** [comment]
@genie-omr build all

File: compiler/optimizer/OMRSimplifierHandlers.cpp [hunk 3]
```diff
@@ -4326,9 +4326,7 @@ static void unsignedIntCompareNarrower(TR::Node * node, TR::Simplifier * s, TR::
       TR::ILOpCodes secondOp = secondChild->getOpCodeValue();
       if (firstOp == TR::su2i && firstChild->getReferenceCount()==1)
          {
-         if (secondOp == TR::su2i ||
-             (secondOp == TR::iuconst               &&
-              secondChild->getUnsignedInt() <= USHRT_MAX))
+         if (secondOp == TR::su2i || secondChild->getUnsignedInt() <= USHRT_MAX)
```
Comment (vijaysun-omr) [inline compiler/optimizer/OMRSimplifierHandlers.cpp]: Is this logically equivalent to what was there before the change ? It seems to me that you should have omitted the ```secondChild->getUnsignedInt() <= USHRT_MAX``` test on the basis that ```secondOp == TR::iuconst``` should be false.

Comment (wbh123456) [reply compiler/optimizer/OMRSimplifierHandlers.cpp]: Yes, this was a mistake. `(secondOp == TR::iuconst&&secondChild->getUnsignedInt() <= USHRT_MAX)` should all be omitted as it will always be `false`. Fix in the new force-push.

File: compiler/x/codegen/BinaryEvaluator.cpp [hunk 4]
```diff
@@ -1252,8 +1252,7 @@ TR::Register *OMR::X86::TreeEvaluator::caddEvaluator(TR::Node *node, TR::CodeGen
...
-       testRegister == NULL         &&
+   if (testRegister == NULL         &&
```
Comment (vijaysun-omr) [inline compiler/x/codegen/BinaryEvaluator.cpp]: Is this logic what you intended here ? I ask because the opcode value not being ```cconst``` suggests that this if condition ought not to be entered.

Comment (wbh123456) [reply compiler/x/codegen/BinaryEvaluator.cpp]: I believe you are reviewing the outdated version of this. I have deleted this whole if statement.

**Comment by wbh123456** [comment]
Did some code simplifying in `x/codegen/BinaryEvaluator.cpp` in the latest [commit].(https://github.com/eclipse/omr/pull/4033/commits/0932037936ea4310a07be2740c897d1435ae1fbf)

**Comment by vijaysun-omr** [comment]
@genie-omr build all

**Comment by vijaysun-omr** [comment]
@andrewcraik are you okay with this PR ?

**Comment by andrewcraik** [comment]
I don't think we should merge this yet. OpenJ9 has at least one place that can still generate `cconst` nodes (https://github.com/eclipse/openj9/blob/3306c99a7dfcc5b033e98dad7ff810a3e546b3c3/runtime/compiler/optimizer/SequentialStoreSimplifier.cpp#L1777) and other downstream projects may have some as well. Until downstream projects have cleaned up their use this change would break them by removing the evaluators they are likely using and optimizations they may be relying upon. We should help those projects clean-up their usage before we remove the opcode from OMR.

**Comment by vijaysun-omr** [comment]
@wbh123456 ping ?

**Comment by wbh123456** [comment]
> @wbh123456 ping ?

Still working on removing unsigned opcodes from OpenJ9. Once that's done, we can merge this.

**Comment by vijaysun-omr** [comment]
@wbh123456 any update on this ?

**Comment by wbh123456** [comment]
> @wbh123456 any update on this ?

This PR is still work in progress as it we need to stop generating all `const` IL Opcode before we can remove the references to them.  https://github.com/eclipse/omr/pull/4347.

**Comment by wbh123456** [comment]
There is one more reference to `Node::cconst()` in OpenJ9. It is going to be removed in https://github.com/eclipse/openj9/pull/7909. This PR will be ready once that PR is merged.

**Comment by wbh123456** [comment]
This PR is ready for a review. @andrewcraik @vijaysun-omr

**Comment by Leonardo2718** [comment]
@genie-omr build all

**Comment by vijaysun-omr** [comment]
The logic after the cleanup looks right to me. Given that the downstream project dependencies have now been eliminated, I think this can go in if @andrewcraik gives his approval.

File: compiler/optimizer/OMRSimplifierHelpers.cpp [hunk 5]
```diff
@@ -346,8 +346,6 @@ TR::Node *foldRedundantAND(TR::Node * node, TR::ILOpCodes andOpCode, TR::ILOpCod
       {
       switch(constOpCode)
          {
-         case TR::sconst: case TR::cconst:
```
Comment (andrewcraik) [inline compiler/optimizer/OMRSimplifierHelpers.cpp]: I don't think this is right - didn't you mean to just delete the `cconst` case - the sconst is not being removed right?

File: compiler/x/codegen/BinaryEvaluator.cpp (hunk 2 above)
Comment (andrewcraik) [reply compiler/x/codegen/BinaryEvaluator.cpp]: Yeah I think this should be sconst with appropriate updates - I don't think the whole case can be removed

File: compiler/x/codegen/BinaryEvaluator.cpp (hunk 1 above)
Comment (andrewcraik) [reply compiler/x/codegen/BinaryEvaluator.cpp]: I think this may need to be changed to sconst - I don't think it can be completely removed.

**Comment by andrewcraik** [comment]
There are two removals in the codegen that don't seem right - ccsont should be mapped to sconst for these I believe @wbh123456 please review and comment. Once we conclude this discussion I think this will be ready.

File: compiler/optimizer/OMRSimplifierHelpers.cpp (hunk 5 above)
Comment (wbh123456) [reply compiler/optimizer/OMRSimplifierHelpers.cpp]: Sorry, that was a mistake, only `cconst` should be deleted. Fixed in c215804

File: compiler/x/codegen/BinaryEvaluator.cpp (hunk 1 above)
Comment (wbh123456) [reply compiler/x/codegen/BinaryEvaluator.cpp]: Fixed in c215804

File: compiler/x/codegen/BinaryEvaluator.cpp (hunk 2 above)
Comment (wbh123456) [reply compiler/x/codegen/BinaryEvaluator.cpp]: Fixed in c215804.

**Comment by wbh123456** [comment]
> There are two removals in the codegen that don't seem right - ccsont should be mapped to sconst for these I believe @wbh123456 please review and comment. Once we conclude this discussion I think this will be ready.

I agree that those two removals are not correct. I have reverted back the if statement and mapped `cconst` to `sconst` in c215804. I don't think there are anything else needed to be updated, do you agree on that?

**Comment by vijaysun-omr** [comment]
@genie-omr build all

File: compiler/optimizer/OMRSimplifierHandlers.cpp [hunk 6]
```diff
@@ -4340,24 +4338,6 @@ static void unsignedIntCompareNarrower(TR::Node * node, TR::Simplifier * s, TR::
                   dumpOptDetails(s->comp(), "Integer Compare Narrower: found both children c2i in method %s\n", s->comp()->signature());
                   }
                }
-            else if (secondOp == TR::iuconst)
```
Comment (andrewcraik) [inline compiler/optimizer/OMRSimplifierHandlers.cpp]: Why is this removed? shouldn't it be an iconst?

File: compiler/optimizer/OMRSimplifierHandlers.cpp [hunk 7]
```diff
@@ -4408,9 +4388,7 @@ static void unsignedIntCompareNarrower(TR::Node * node, TR::Simplifier * s, TR::
          }
       else if (firstOp == TR::b2i && firstChild->getReferenceCount() == 1)
          {
-         if (secondOp == TR::b2i ||
-             (secondOp == TR::iuconst               &&
```
Comment (andrewcraik) [inline compiler/optimizer/OMRSimplifierHandlers.cpp]: shouldn't this be an iconst with a getUnsignedInt to test against SCHAR_MAX?

File: compiler/optimizer/OMRSimplifierHandlers.cpp [hunk 8]
```diff
@@ -4424,25 +4402,6 @@ static void unsignedIntCompareNarrower(TR::Node * node, TR::Simplifier * s, TR::
                   dumpOptDetails(s->comp(), "Integer Compare Narrower: found both children b2i in method %s\n", s->comp()->signature());
                   }
                }
-            else if (secondOp == TR::iuconst)
```
Comment (andrewcraik) [inline compiler/optimizer/OMRSimplifierHandlers.cpp]: again iconst with appropriate adaptations rather than removal...

File: compiler/optimizer/OMRSimplifierHandlers.cpp (hunk 6 above)
Comment (wbh123456) [reply compiler/optimizer/OMRSimplifierHandlers.cpp]: This is a mistake. The `TR::iuconst` in this `IntComp` pattern matching should be mapped to `TR::iconst` in this case. I have fixed this problem in cdd352b

**Comment by wbh123456** [comment]
@andrewcraik I have fixed the false removal of `TR::iuconst` in cdd352b. I have mapped the `TR::iuconst` to `TR::iconst`. In addition, I have also reverted back the test against `SCHAR_MAX` as it is still valid in with `TR:iconst`

File: compiler/optimizer/OMRSimplifierHandlers.cpp (hunk 7 above)
Comment (wbh123456) [reply compiler/optimizer/OMRSimplifierHandlers.cpp]: Fixed.

File: compiler/optimizer/OMRSimplifierHandlers.cpp (hunk 8 above)
Comment (wbh123456) [reply compiler/optimizer/OMRSimplifierHandlers.cpp]: Fixed

**Comment by wbh123456** [comment]
@andrewcraik I have fixed `iuconst` mapping problems in `OMRSimplifierHandlers.cpp`.  I have also scanned all the mappings of deprecated unsigned const opcodes, which all look appropriate to me.  This PR is ready for another review.

File: compiler/optimizer/VPHandlers.cpp [hunk 9]
```diff
@@ -10748,7 +10748,7 @@ static TR::Node *constrainCmplessthan(OMR::ValuePropagation *vp, TR::Node *node,
           performTransformation(vp->comp(), "%sChanging node [%p] %s into constant %d\n", OPT_DETAILS, node, node->getOpCode().getName(), result))
          {
          vp->removeChildren(node);
-         TR::ILOpCodes op = /*isUnsigned ? TR::iuconst : */TR::iconst;
+         TR::ILOpCodes op = TR::iconst;
```
Comment (andrewcraik) [inline compiler/optimizer/VPHandlers.cpp]: can you just put iconst directly into the recreate to save a needless local?

File: compiler/x/codegen/BinaryEvaluator.cpp [hunk 10]
```diff
@@ -1252,49 +1252,49 @@ TR::Register *OMR::X86::TreeEvaluator::caddEvaluator(TR::Node *node, TR::CodeGen
...
+      TR::MemoryReference  *tempMR = generateX86MemoryReference(targetRegister, value, cg);
+      targetRegister = cg->allocateRegister();
+      generateRegMemInstruction(LEA4RegMem, node, targetRegister, tempMR, cg);
+      }
+   else
+      {
+      if (value >= 0 && value <= 127)
          {
-         if (value >= 0 && value <= 127)
+         if (value == 1)
```
Comment (andrewcraik) [inline compiler/x/codegen/BinaryEvaluator.cpp]: I'm not sure this is right - the sconst is read as unsigned - isn't it possible for the value to be between 128 and 255 inclusive. I know the node is now sconst but that doesn't change the range when the const is still read as a uint16_t...

File: compiler/x/codegen/BinaryEvaluator.cpp [hunk 11]
```diff
@@ -1252,49 +1252,49 @@ TR::Register *OMR::X86::TreeEvaluator::caddEvaluator(TR::Node *node, TR::CodeGen
       }
 
    TR::Register * testRegister = secondChild->getRegister();
-   if (secondChild->getOpCodeValue() == TR::cconst &&
-       testRegister == NULL         &&
-       performTransformation(comp, "O^O CaddEvaluator: checking that the store has not happened yet. Target register:  %x\n", testRegister))
+   if (secondChild->getOpCodeValue() == TR::sconst &&
+      testRegister == NULL         &&
+      performTransformation(comp, "O^O CaddEvaluator: checking that the store has not happened yet. Target register:  %x\n", testRegister))
+   {
```
Comment (andrewcraik) [inline compiler/x/codegen/BinaryEvaluator.cpp]: Something weird happened here in the formatting - can you sort the diff to make more sense - looks like the if body got outdented?

File: compiler/z/codegen/UnaryEvaluator.cpp [hunk 12]
```diff
@@ -117,13 +117,8 @@ TR::Register *
...
    {
    TR::Register * tempReg = node->setRegister(cg->allocateRegister());
-   if (node->getOpCodeValue() == TR::buconst)
-      generateLoad32BitConstant(cg, node, node->getByte() & 0xFF, tempReg, true);
-   else
-      {
-      TR_ASSERT(1, "Should not have bconst node in the final il-tree!\n");
-      generateLoad32BitConstant(cg, node, node->getByte(), tempReg, true);
-      }
+   TR_ASSERT(1, "Should not have bconst node in the final il-tree!\n");
```
Comment (andrewcraik) [inline compiler/z/codegen/UnaryEvaluator.cpp]: @fjeremic is this assert even valid? this should be removed shouldn't it (in the context of OMR)

**Comment by andrewcraik** [comment]
Opcode transform and handling now looks right - just a few smaller things to look at and we can look to merge this.

File: compiler/optimizer/VPHandlers.cpp (hunk 9 above)
Comment (wbh123456) [reply compiler/optimizer/VPHandlers.cpp]: Fixed in e4cb262

File: compiler/x/codegen/BinaryEvaluator.cpp (hunk 11 above)
Comment (wbh123456) [reply compiler/x/codegen/BinaryEvaluator.cpp]: I am not sure what happens here. The formatting looks exactly the same on both sides. Git diff tool on my local repo also can't detect a difference except from line `1255`.  Lines from `1256` to `1297` should not have changed at all. I have tried reverted the change in line `1255`, it did not work.

File: compiler/z/codegen/UnaryEvaluator.cpp (hunk 12 above)
Comment (wbh123456) [reply compiler/z/codegen/UnaryEvaluator.cpp]: @fjeremic Ping

Comment (fjeremic) [reply compiler/z/codegen/UnaryEvaluator.cpp]: This assert seems bogus. `bconst` trees should be valid in the OMR context. I suspect this may be some Java-ism leaking over. This evaluator looks correct. We should remove the assert.

Comment (wbh123456) [reply compiler/z/codegen/UnaryEvaluator.cpp]: I have deleted the assert in c56d3c5.

File: compiler/optimizer/LoopReducer.cpp [hunk 13]
```diff
@@ -1811,7 +1811,7 @@ TR_Arraytranslate::checkBreak(TR::Block * breakBlock, TR::Node * breakNode, TR::
       //      we also conservatively ensure the range is inside the represented ranges such that we can
       //      add or subtract 1 from the value and still represent it in 2**16 bits for a default value
       //      if we need to generate a table.
-      dumpOptDetails(comp(), "...break tree does not have bconst/cconst/iconst, or not in range - no arraytranslate reduction\n");
+      dumpOptDetails(comp(), "...break tree does not have bconst/iconst, or not in range - no arraytranslate reduction\n");
```
Comment (andrewcraik) [inline compiler/optimizer/LoopReducer.cpp]: shouldn't sconst be in here rather than deleting cconst?

Comment (wbh123456) [reply compiler/optimizer/LoopReducer.cpp]: The if statement is not checking `cconst` nor `bconst`. Maybe this `dumpOptDetails()` is not up to date. Should I just delete both `cconst` and `bconst`?

Comment (andrewcraik) [reply compiler/optimizer/LoopReducer.cpp]: LoopReducer is not heavily used anymore so I would say if the condition is only checking for iconst just update the details to say as much

Comment (wbh123456) [reply compiler/optimizer/LoopReducer.cpp]: Fixed.

**Comment by 0xdaryl** [comment]
@genie-omr build all
